3주차 Spring MVC 테스트 과제

kckoh opened the pull request:
안녕하세요.
아직 전부다 테스트 케이스를 전부 구현하지 못했습니다
이번주도 열심히 해보겠습니다.

kckoh committed:
3주차 Spring MVC 테스트 과제

안녕하세요.
아직 전부다 테스트 케이스를 전부 구현하지 못했습니다
이번주도 열심히 해보겠습니다.

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerTest.java:
```diff
@@ -0,0 +1,87 @@
+    @Test
+    void detailWithValidId() {
+        //given
+        Task task = new Task();
+        task.setTitle("title");
+        task.setId(1L);
```
`@BeforeEach`에 있는 코드 같습니다!

kckoh replied:
리펙토링 하였습니다!

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerTest.java:
```diff
@@ -0,0 +1,87 @@
+        taskController = new TaskController(taskService);
+        task = new Task();
+        task.setTitle("title");
+        task.setId(1L);
```
Task라는 객체를 초기화하기 위해선 항상 N개의 줄이 필요하군요! 한줄로 만들어져야 더욱 안전한 코드가 됩니다!

Effective Java의 2번을 참고하세요! 그리고 그곳에서 왜 setter를 사용한 방식이 위험한지를 찾아 이곳에 공유해주세요!

kckoh replied:
안전한 코드를 만들기 위해서 Task 클라스를 불변 객채로 만들었습니다. setter를 지우고 constructor와 final 키워드로 구현 하였습니다.

ssisksl77 replied:
좋습니다!

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerTest.java:
```diff
@@ -0,0 +1,87 @@
+
+    }
+
+    @Test
+    void create() {
```
테스트에 설명이 좀 더 있으면 좋겠어요!

kckoh replied:
혹시 comment 외 테스트 설명을 추가할때 `@DisplayName` annotation을 쓰면 될까요? 다른 annotation이 있을까요?

ssisksl77 replied:
`@DisplayName` 으로 완벽합니다!

kckoh committed:
refactor: remove setTitle setId from Task

고친점:
    Task 클라스에 있는 세터를 지우고 task 클라스를 불변으로 만들었습니다.
    taskService에 있는 setter를 리펙토링 하였습니다
    setter를 쓰는 Test 케이스들을 리펙토링 하였습니다

kckoh committed:
add: TaskTest implemented

Task 모델 테스트 케이스 작성

kckoh committed:
add: implement TaskServiceTest

고친점:
    TaskService 테스트 작성

kckoh committed:
add: implement TaskControllerTest

고친점:
    TaskController 유닛 테스트 작성

kckoh committed:
fix: trying to fix TaskControllerWebTest

아직 BDD 개념이 이해가 안가는데 혹시 좋은 resource 알려주실수 있으세요?

kckoh commented:
아직 bdd 개념과 given()을 어떻게 쓰는지 햇갈리는데 혹시 설명 해줄수 있으실까요? given()이 어떤 동작을 하고 왜 쓰는지 아직도 이해가 안가네요 ㅠㅠ

ssisksl77 commented:
given은 when이 시작되기 전의 상태를 말합니다. 여기서 상태란 내가 어떤 행위를 하기 전에 존재(준비)되어야하는 환경을 말합니다!

따라서 어떤 상태(환경)이 주어지고(given) 
내가 어떤 일하면(when)
무엇을 리턴한다(then)

이런식이 될 것입니다.

즉, 테스트를 할 때 어떤 환경인지를 나열하기 위해 given이라는 단계가 필요하죠!

kckoh committed:
add: implement TaskControllerWebTest

고친점:
    TaskControllerWebTest 테스트 구현
    테스트 커버리지 100%

ssisksl77 commented:
모든 테스트에 `@DiaplayName` 어노테이션으로 설명을 붙이는 것이 선행되어야겠습니다!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/models/Task.java:
```diff
@@ -1,23 +1,25 @@
+    }
 
+
+    public Task updateTitle(String title) {
+        return new Task(this.id, title);
+    }
```
불변성을 도모하는 코드군요!

ssisksl77 replied:
변경된 Task를 리턴하는 것은 자명하지만 그래도 어떤 일을 하고 무엇을 리턴하는지 주석에 적는 것이 좋겠어요!

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,109 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.catchThrowable;
+
+class TaskServiceTest {
```
한번 계층형 BDD스타일을 적용해보는 것도 큰 도움이 될 거에요!

https://johngrib.github.io/wiki/junit5-nested/

kckoh committed:
refactor: apply Nested BDD to HelloControllerMockMVCTest HelloControllerTest Task

고친점:
    HelloControllerMockMVCTest HelloControllerTest Nested BDD를 적용 하였니다.

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/models/TaskTest.java:
```diff
@@ -0,0 +1,59 @@
+        @DisplayName("if an updated task is given")
+        @Nested
+        class Context_with_update_title {
+            private Task task;
+
+            @BeforeEach
+            void setup() {
+                task = new Task(1L, "title");
+            }
```
```suggestion
           private final Task task = new Task(1L, "title");
```

Task가 불변이고 계속 상태라 바꿔서 사용하지 않을 것이라면 그대로 써도 되지 않을까해요!

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/models/TaskTest.java:
```diff
@@ -0,0 +1,59 @@
+    @DisplayName("UpdateTitle Method")
+    @Nested
+    class Describe_updateTitle {
+        @DisplayName("if an updated task is given")
```
DisplayName을 한글보다 영어를 선호하신 이유가 궁금합니다!

kckoh replied:
지금 캐나다에서 공부하고 있고 취업도 캐나다에서 할 생각이라서 코멘트나 `Displayname` 을 전부 영어로 작성하고 있습니다.

kckoh replied:
영어로 적었을 때 예를 들어서 returns/ throws 쓰면 코드랑 코멘트가 일치해서 좀 더 보기 쉽겠다고 생각합니다. 그리고  작업 환경 변경 시  인코딩 이슈가 있을 수 있다고 생각해서 모든 코멘트는 영어로 작성하고 있습니다.

ssisksl77 replied:
아하! 좋습니다! 협업하는 팀원들이 영어주석에 다들 익숙하다면 좋은 선택일 수 있겠어요!

하지만 함께하는 팀원 중에 영어에 불편함이 있다면 한글이 적절할 것 같아요!

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/models/TaskTest.java:
```diff
@@ -0,0 +1,59 @@
+
+            @BeforeEach
+            void setup() {
+                task = new Task(1L, "title");
```
```suggestion
                task = new Task(id1, "title");
```

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/application/TaskService.java:
```diff
@@ -24,30 +24,23 @@ public Task getTask(Long id) {
     }
 
     public Task deleteTask(Long id) {
         Task task = getTask(id);
         tasks.remove(task);
-
         return task;
     }
```
`task` 를 못찾으면 null 리턴하는걸까요?

kckoh committed:
Update app/src/test/java/com/codesoom/assignment/models/TaskTest.java

refactor to immutable object

Co-authored-by: 남영환 <ssisksl77@gmail.com>

kckoh committed:
Update app/src/test/java/com/codesoom/assignment/models/TaskTest.java

refactor to id1

Co-authored-by: 남영환 <ssisksl77@gmail.com>

kckoh committed:
refactor: apply D-C-I to TaskControllerTest class

고친점:
    TaskControllerTest를 D-C-I 패턴으로 리팩토링 하였습니다.

kckoh committed:
Merge branch 'kckoh' of https://github.com/kckoh/spring-week3-assignment-1 into kckoh

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/models/TaskTest.java:
```diff
@@ -0,0 +1,57 @@
+        @DisplayName("if an updated task is given")
+        @Nested
+        class Context_with_update_title {
+<<<<<<< HEAD
+            private Task task = new Task(1L, "title");
+=======
+           private final Task task = new Task(1L, "title");
+>>>>>>> db72f78c24c5968d9b97c5262a058827dbdbc546
```
해결을 해야할 충돌이 그대로 있군요. PR에 올릴 때는 로컬 테스트가 성공한 이후에 올리는 습관을 들이는 것이 좋습니다!

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerTest.java:
```diff
@@ -0,0 +1,184 @@
+
+
+
+//    @Test
+//    @DisplayName("add two tasks and checks the size")
+//    void createTwoTasks() {
+//        Task task1 = new Task(1L, "title1");
+//        taskController.create(task1);
+//        Task task2 = new Task(2L, "title2");
+//        taskController.create(task2);
+//
+//        assertThat(taskController.list()).hasSize(2);
+//    }
+//
+//
+//
```
주석을 처리했음에도 PR에 함께 있다면 그 이유가 있어야합니다!

kckoh committed:
fix: return delete() returns optional in TaskService

고친점:
    delete() 리턴 타입을 optional<task>로 변경

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/application/TaskService.java:
```diff
@@ -34,10 +35,10 @@ public Task updateTask(Long id, Task source) {
 
-    public Task deleteTask(Long id) {
+    public Optional<Task> deleteTask(Long id) {
         Task task = getTask(id);
         tasks.remove(task);
-        return task;
+        return Optional.ofNullable(task);
     }
```
같은 위치에 있는 `getTask` 는 task가 존재했을 경우 예외를 던졌습니다.

deleteTask에는 왜 Optional을 선택하셨는지 궁금합니다!

kckoh replied:
null이 나올수도 있겟다고 생각해서 optional로 했습니다

ssisksl77 commented:
한주간 고생하셨습니다!

kckoh commented:
회고가 늦었네요 죄송합니다.
https://kckoh.hashnode.dev/2-oct-2022-retrospective
